File: android/app/src/main/java/info/bagen/rust/plaoc/microService/sys/nativeui/helper/QueryHelper.kt
package info.bagen.rust.plaoc.microService.sys.nativeui.helper

import androidx.compose.foundation.layout.WindowInsets
import androidx.compose.ui.graphics.Color
import info.bagen.rust.plaoc.microService.core.NativeMicroModule
import info.bagen.rust.plaoc.microService.helper.ColorJson
import info.bagen.rust.plaoc.microService.helper.gson
import info.bagen.rust.plaoc.microService.helper.toJsonAble
import info.bagen.rust.plaoc.microService.sys.nativeui.navigationBar.NavigationBarController
import info.bagen.rust.plaoc.microService.sys.nativeui.safeArea.SafeAreaController
import info.bagen.rust.plaoc.microService.sys.nativeui.statusBar.StatusBarController
import info.bagen.rust.plaoc.microService.sys.nativeui.virtualKeyboard.VirtualKeyboardController
import org.http4k.core.Request
import org.http4k.lens.Query
import org.http4k.lens.boolean
import org.http4k.lens.string

class QueryHelper {
    companion object {
        init {
            NativeMicroModule.ResponseRegistry.registryJsonAble(Color::class.java) { it.toJsonAble() }
            NativeMicroModule.ResponseRegistry.registryJsonAble(WindowInsets::class.java) { it.toJsonAble() }
            NativeMicroModule.ResponseRegistry.registryJsonAble(StatusBarController::class.java) { it.toJsonAble() }
            NativeMicroModule.ResponseRegistry.registryJsonAble(NavigationBarController::class.java) { it.toJsonAble() }
            NativeMicroModule.ResponseRegistry.registryJsonAble(SafeAreaController::class.java) { it.toJsonAble() }
            NativeMicroModule.ResponseRegistry.registryJsonAble(VirtualKeyboardController::class.java) { it.toJsonAble() }
        }

        fun init() {
            // 确保 init 里头的类被注册
        }

        val query_color = Query.string().optional("color")
        val query_style = Query.string().optional("style")
        val query_visible = Query.boolean().optional("visible")
        val query_overlay = Query.boolean().optional("overlay")
        inline fun color(req: Request) = query_color(req)?.let {
            gson.fromJson(it, ColorJson::class.java).toColor()
        }

        inline fun style(req: Request) = query_style(req)?.let {
            BarStyle.from(it)
        }

        inline fun visible(req: Request) = query_visible(req)
        inline fun overlay(req: Request) = query_overlay(req)
    }
}

File: next/dweb-browser/src/Platforms/iOS/MicroModule/Browser/NativeUI/Helper/QueryHelper.cs
using System.Text.Json;
using DwebBrowser.MicroService.Browser.NativeUI.Base;
using DwebBrowser.MicroService.Browser.NativeUI.SafeArea;
using DwebBrowser.MicroService.Browser.NativeUI.StatusBar;
using DwebBrowser.MicroService.Browser.NativeUI.NavigationBar;
using DwebBrowser.MicroService.Browser.NativeUI.VirtualKeyboard;
using DwebBrowser.MicroService.Http;

#nullable enable

namespace DwebBrowser.Helper;

public class QueryHelper
{
    static QueryHelper()
    {
        ResponseRegistry.RegistryJsonAble<StatusBarController>(
            typeof(StatusBarController), it => it.GetState());
        ResponseRegistry.RegistryJsonAble<NavigationBarController>(
            typeof(NavigationBarController), it => it.GetState());
        ResponseRegistry.RegistryJsonAble<VirtualKeyboardController>(
            typeof(VirtualKeyboardController), it => it.GetState());
        ResponseRegistry.RegistryJsonAble<SafeAreaController>(
            typeof(SafeAreaController), it => it.GetState());
    }

}
public static class PureRequestQueryExtensions
{

    public static ColorJson? QueryColor(this PureRequest req, string key) => req.ParsedUrl?.SearchParams.Get(key)?.Let(it =>
    {
        return JsonSerializer.Deserialize<ColorJson>(it)!;
    });

    public static BarStyle? QueryStyle(this PureRequest req, string key) => req.ParsedUrl?.SearchParams.Get(key)?.Let(it =>
    {
        return new BarStyle(it.ToUpper());
    });

    public static bool? QueryBool(this PureRequest req, string key) => req.ParsedUrl?.SearchParams.Get(key)?.ToBooleanStrictOrNull();
    public static string? QueryString(this PureRequest req, string key) => req.ParsedUrl?.SearchParams.Get(key);
    public static string QueryStringRequired(this PureRequest req, string key) => QueryString(req, key) ?? throw new KeyNotFoundException("require query." + key);
}
